### PR TITLE
Support $elemMatch array query operator

### LINKED ISSUE #699: Support `$elemMatch` array query operator

### PR DESCRIPTION
**Pull request by w84thesun**
Closes #699.

### CODE REVIEW

**Comment by w84thesun** [comment]
I would really appreciate any test case suggestions. 😄

**Comment by seeforschauer** [comment]
> I would really appreciate any test case suggestions. smile

* What if there are two `$elemMatch` inside the filter?
* what if there are two similar operates inside one `$elemMatch`?
* Please see error `ErrElemMatchExclusionInInclusion` and ErrElemMatchInclusionInExclusion in #551 these errors also applicable to `$elemMatch` operator?
* `filterFieldExpr` does almost all stuff, And what if the expression field contains `$and` `$or` `$nor` how does it behave? 
* please add `$elemMatch cannot contain $where expression`

**Review by seeforschauer — CHANGES_REQUESTED** [review]
LGTM and some more tests are required.

File: internal/handlers/common/filter.go [hunk 1]
```diff
@@ -495,6 +495,13 @@ func filterFieldExpr(doc *types.Document, filterKey string, expr *types.Document
 				return false, err
 			}
 
+		case "$elemMatch":
```
Comment (rumyantseva) [inline internal/handlers/common/filter.go:498]: Should `$elemMatch` be before `$size` in this list?

File: integration/query_array_test.go [hunk 2]
```diff
@@ -212,6 +212,90 @@ func TestQueryArrayDotNotation(t *testing.T) {
...
+			assert.Equal(t, tc.expectedIDs, CollectIDs(t, actual))
+		})
+	}
+}
+
+func TestQueryElemMatchOperator(t *testing.T) {
+	t.Parallel()
+	ctx, collection := setup(t, shareddata.Scalars, shareddata.Composites)
+
+	for name, tc := range map[string]struct {
```
Comment (rumyantseva) [inline integration/query_array_test.go:233]: All the test cases have `gt` or `lt`. Can we have one that is a mix of conditions, something like `{ value: { $elemMatch: { field1: "a", field2: { $gt: 42 } } } }`?

Also, it seems that `$ne` might give interesting cases, and we should add it to the tests.

File: internal/handlers/common/filter.go [hunk 3]
```diff
@@ -1084,3 +1091,20 @@ func filterFieldValueByTypeCode(fieldValue any, code typeCode) (bool, error) {
 
 	return true, nil
 }
+
+// filterFieldExprElemMatch handles {field: {$elemMatch: value}}.
+// Returns false if doc value is not an array.
+func filterFieldExprElemMatch(doc *types.Document, filterKey string, exprValue any) (bool, error) {
```
Comment (rumyantseva) [inline internal/handlers/common/filter.go:1097]: Should we have anything special about `$text` and `$where` not being supported in `$elemMatch`?

**Review by rumyantseva — CHANGES_REQUESTED** [review]
A few suggestions of what could be added.

File: internal/handlers/common/filter.go (hunk 1 above)
Comment (w84thesun) [reply internal/handlers/common/filter.go]: Moved it.

File: internal/handlers/common/filter.go (hunk 3 above)
Comment (w84thesun) [reply internal/handlers/common/filter.go:1100]: Added correct response for `$text` and `$where`.

**Comment by w84thesun** [comment]
I suggest moving `$ne`, `$not`, and `$and` out of scope as it seems that their implementation would drastically increase PR size.

File: integration/query_array_test.go (hunk 2 above)
Comment (w84thesun) [reply integration/query_array_test.go:233]: I've added a similar case to that one you suggest → `GtField`.
Also please see https://github.com/FerretDB/FerretDB/pull/707#issuecomment-1152037874

**Comment by w84thesun** [comment]
> * What if there are two `$elemMatch` inside the filter?
> * What if there are two similar operators inside one `$elemMatch`?
> * Please see error `ErrElemMatchExclusionInInclusion` and `ErrElemMatchInclusionInExclusion` in [Support `$elemMatch`  projection basic comparison query operators #551](https://github.com/FerretDB/FerretDB/pull/551). Are these errors for `$elemMatch` query applicable to `$elemMatch` operator?
> * `filterFieldExpr` does almost all stuff. And what if the expression field contains
>   
>   * `$and`
>   * `$or`
>   * `$nor`
>   * `$not` - the result of the query must be similar to the result without `$elemMatch`
>   * `$ne` - the result of the query must be similar to the result without `$elemMatch`
>     How does it behave?
> * `$elemMatch` cannot contain `$where`, could you please add a test for it?
> * `$elemMatch` cannot contain `$text`, could you please add a test for it?
> * WDYT of adding the test where an operator and a scalar field expression are present?  `bson.D{{"value", bson.D{{"$elemMatch", bson.D{{"$gt", int32(0)}, {"somefield", int32(42)}}}}}}`

Thanks for the suggestions!

Two `$elemMatch` or duplicate `$gt` are out of scope → #364 
I've added similar query operators → `GtLt` case
Looks like errors are not applicable.
For `$and` etc. please see https://github.com/FerretDB/FerretDB/pull/707#issuecomment-1152037874
Errors for `$where` and `$text` where added.
Case `GtField` added.

**Comment by seeforschauer** [comment]
> I suggest moving `$ne`, `$not`, `$and`, `$or` and `$nor` out of scope as it seems that their implementation would drastically increase PR size.

@w84thesun would you please add an issue link or create on if it doesn't exists?

File: internal/handlers/common/filter.go [hunk 4]
```diff
@@ -1084,3 +1093,35 @@ func filterFieldValueByTypeCode(fieldValue any, code typeCode) (bool, error) {
...
+	expr, ok := exprValue.(*types.Document)
+	if !ok {
+		return false, NewErrorMsg(ErrBadValue, "$elemMatch needs an Object")
+	}
+
+	for _, key := range expr.Keys() {
+		if slices.Contains([]string{"$text", "$where"}, key) {
+			return false, NewErrorMsg(ErrBadValue, fmt.Sprintf("%s can only be applied to the top-level document", key))
+		}
+
```
Comment (seeforschauer) [inline internal/handlers/common/filter.go:1116]: Would you please add a TODO with the issue number here? Thank you.

Comment (w84thesun) [reply internal/handlers/common/filter.go:1116]: Added

**Review by seeforschauer — APPROVED** [review]

**Review by rumyantseva — APPROVED** [review]
LGTM. I agree with the idea of limiting the scope of this task and creating a separate issue for `$ne, $not, $and, $or, $nor`.

**Review by AlekSi — APPROVED** [review]

**Pull request merged**